### ISSUE TITLE
Instead of Boolean, why not a List of Strings? Or a Map of Strings to Frequencies?

### ISSUE THREAD

**Issue by ashnur** [dubzzz/fast-check#227]
https://github.com/dubzzz/fast-check/blob/46fea0eefb2f8faa59bb6027c6d1ee90119d769e/src/check/arbitrary/RecordArbitrary.ts#L72-L73

`[string]`:
```
 for (const k of Object.keys(recordModel))
    updatedRecordModel[k] = option((recordModel as { [key: string]: Arbitrary<any> })[k].map(v => ({ value: v })), constraints.requiredKeys.includes(k) ? 1 : undefined );
```

or

`{string:number}`:
```
for (const k of Object.keys(recordModel))
    updatedRecordModel[k] = option((recordModel as { [key: string]: Arbitrary<any> })[k].map(v => ({ value: v })), constraints.modulatedKeys[k] );
```

**Comment by dubzzz** [comment]
If I well understand the issue, the question is: Why is `withDeletedKeys` a boolean value?

I initially introduced this flag when I started using fast-check against other libs like js-yaml. `fc.record`  proved very useful to check parameters given to a function were really taken into account.

The initial usage was something like:

```js
const paramsArb = fc.record({
  param1: fc.boolean(),
  //... 
});
const instArb = //... 
fc.assert(fc.property(
  instArb, 
  paramsArb, 
  (ins, params) => {
    const out = deserialize(serialize(ins, params));
    assert.deepStrictEqual(ins, out);
  }
));
```

The issue I had with this raw record for `params` was that whenever the framework found an issue it sent me back a `params` containing all the possible keys inside `params` while only some of them were required to make the test fail.

This is the reason why I introduced the `withDeletedKeys` option.

I believe adding the possibility to choose the keys to delete would be a great idea. We could maybe have:

`withDeletedKeys?: number | string[]`

I think keeping that notion in `withDeletedKeys` would be a good thing - it might be confusing for an end user to see two options doing more or less the same thing :)

**Comment by ashnur** [comment]
personally I like APIs that speak for themselves and are not overloaded. I would much rather have to deal with 3 different options with 3 different behaviors than one key and 3 different behavior. Not to mention how much easier is to search for code using a particular option if the name defines the behavior. 
That said, the ability to fine tune what can get removed from a record is a must for me, I am having large entities that I need to test with lots of different possible keys, but most of these entities have some properties that are "required". 
Another option would be something nicer than .chain() to create new record arbitraries from lots of smaller records, again with varying degree of frequencies. Something like an algorithmic, iterative process that would take a list of records and merge them I am thinking. https://www.npmjs.com/package/object-merge-advanced#markdown-header-optscb

Can I do something to help something like these features get implemented?

**Comment by dubzzz** [comment]
Sorry for the late reply, I just wanted to find a bit more time to answer to your request. 

First of all, it is currently possible to implement such arbitrary quite easily by combining two `record` in one:

```
fc.tuple(
    fc.record({/*required keys*/}),
    fc.record({/*optional keys*/}, {withDeletedKeys: true})
 .map(([o1,o2]) => ({...o1, ...o2));
```

Option 1.

The option to add the ability to provide an array of keys that could be deleted is still valid in my opinion (I agree that it has to be challenged and is maybe not the best way). Some famous javascript tools are relying on such kind of options:
- tslint: configuration accepts either a boolean or an array of string
- babel: targets.node / targets.module...

The idea with this first suggestion would be:
- withDeletedKeys absent: no deleted keys, all keys of the incoming objects will be kept
- withDeletedKeys = false, same 
- withDeletedKeys = true, all the keys of the object are eligible to be removed
- withDeletedKeys = array, only keys in the array are eligible to be removed

Option 2.

Concerning the option of adding `requiredKeys`, `optionalKeys`, `modulatedKeys` or all of them, I see one possible issue: How to handle a configuration having both `withDeletedKeys` and `requiredKeys` (or any other combination of concurrent parameters)?

To answer the question we might maybe take advantage of the type system of typescript to forbid such constructs. But what do we do for JavaScript users?

Option 3.

Another solution might also be to add methods onto the arbitrary produced by record that would allow the user to extend its capabilities. I was thinking as something like: `fc.record({/**/}).withDeletedKeysExcept([/*keys to keep*/])`.

Option 4.

An arbitrary for each case (like `oneof` and `frequency`).

For the moment, all the suggestions are just drafted ideas I had.

**Comment by dubzzz** [comment]
In a way, I am OK that having different keys will be better both for fast-check code and from an user perspective. Meanwhile I do not want to make things implicit: parameters must be usable together.

Implicit things will make documentation more difficult to write, read, understand and people might open issues related to this. 

Let's say we have:
- withDeletedKeys (boolean): truthy = keys can be removed, falsy all keys have to be kept
- requiredKeys (list of strings) : defaulted to empty array, all keys within this array will be kept in the final object - does not delete anything alone, has to be used in conjunction with withDeletedKeys
- deletionFrequency (number): frequency to use whatever the keys that will be considered eligible for deletion - does not delete anything alone

Concerning typings:
- the required keys must be keys of the object, it protects the user from typos (should be possible in ts)
- today when withDeletedKeys is used the produced value moves from `{a: number, b: string}` to `{a: number? , b: string?}`. If `a` is required it should become `{a: number, b: string?}`

Concerning tests, I can give you a hand on this. I actually have an opened issue to simplify them. I have a complex arbitrary helper that check multiple common patterns arbitraries must fulfill:
- same seed, same generated values
- same seed, same shrinks (tested 'randomly' one of the shrinking path)
- no infinite shrinking path, shrinking must reach an end = this one helped me to discover an infinite shrink path that occurred very rarely when I started the project
- values are strictly decreasing along the shrinking path
- ...

**Comment by dubzzz** [comment]
Just an additional question, in my suggestion above, the frequency by key you suggested has been changed into a global frequency for all the keys within the object.

Was there a specific need to have a frequency per key? Is a frequency for all the keys enough?

**Comment by ashnur** [comment]
> Implicit things will make documentation more difficult to write, read, understand and people might open issues related to this.

I agree with not making stuff implicit. So if there is conflict in the options, either has to ignore the least constraining option (maybe even alert the user of this), or somehow avoid configuration that could be conflicting, if this is even possible. 

> Was there a specific need to have a frequency per key? Is a frequency for all the keys enough?

I think it' nice to have the ability to fine tune frequencies by key, however I am unsure how to handle nested objects. 
The use case for this is when generating lots of objects and these vary only slightly in their content. If I just set a global frequency of keys, the resulting objects in general will not look the same as the actual objects created, and I will feel that I am testing something my application will never experience.

**Comment by ashnur** [comment]
what do you think of this? 
https://github.com/dubzzz/fast-check/compare/master...ashnur:master

**Comment by dubzzz** [comment]
@ashnur I just went through the whole thread and your current master branch diff - https://github.com/dubzzz/fast-check/compare/master...ashnur:master

Here are some updates you might have to perform:
```diff
  for (const k of Object.keys(recordModel)) {
    if ( constraints.required && constraints.required.includes(k)  ) {
---      updatedRecordModel[k] = option((recordModel as { [key: string]: Arbitrary<any> })[k].map(v => ({ value: v })));
+++      updatedRecordModel[k] = (recordModel as { [key: string]: Arbitrary<any> })[k].map(v => ({ value: v }));
    } else if ( constraints.frequencies && constraints.frequencies.hasOwnProperty(k)  ) {
      updatedRecordModel[k] = option((recordModel as { [key: string]: Arbitrary<any> })[k].map(v => ({ value: v })), constraints.frequencies[k]);
    } else {
---      updatedRecordModel[k] = option((recordModel as { [key: string]: Arbitrary<any> })[k].map(v => ({ value: v })), 1);
+++      updatedRecordModel[k] = option((recordModel as { [key: string]: Arbitrary<any> })[k].map(v => ({ value: v })));
    }
  }
```

Not sure for this one but it seems that if there is no `withDeletedKeys` flag toggled we can safely say that the arbitrary will be a simple `Arbitrary<T>` with no deleted keys.

```diff
---type ConstrainedArbitrary<T, Constraints> = Constraints extends DeletedKeys | DeletedKeysDepreciated | DeleteConditions
+++type ConstrainedArbitrary<T, Constraints> = Constraints extends DeletedKeys | DeletedKeysDepreciated
```

Another improvement that we might bring later would be not to mark the keys set in `required` as optional in:


```ts
type ConstrainedArbitrary<T, Constraints> = Constraints extends DeletedKeys | DeletedKeysDepreciated
  ? Arbitrary<Partial<T>> // <-- here but I haven't checked if it was even feasible
  : Arbitrary<T>;
```

**Comment by ashnur** [comment]
Typescript is a blocker for me.
